From: Daniel-Constantin Mierla <miconda@gmail.com>
Date: Fri, 21 Apr 2017 21:39:22 +0200
Subject: dialog: more safety checks to detect if dialog is gone after event
 route execution

- avoid execution of script callbacks for event route
- related to GH #1059 and #1069
---
 modules/dialog/dlg_handlers.c | 36 +++++++++++++++++++-----------------
 modules/dialog/dlg_hash.h     |  2 +-
 modules/dialog/dlg_profile.c  |  4 ++++
 modules/dialog/dlg_var.c      | 12 ++++++++++++
 modules/ims_dialog/dlg_var.c  |  3 +++
 5 files changed, 39 insertions(+), 18 deletions(-)

diff --git a/modules/dialog/dlg_handlers.c b/modules/dialog/dlg_handlers.c
index 6b00d7d..7fc6347 100644
--- a/modules/dialog/dlg_handlers.c
+++ b/modules/dialog/dlg_handlers.c
@@ -455,7 +455,11 @@ static void dlg_onreply(struct cell* t, int type, struct tmcb_params *param)
 		event = DLG_EVENT_RPL3xx;
 
 	next_state_dlg( dlg, event, &old_state, &new_state, &unref);
-	dlg_run_event_route(dlg, (rpl==FAKED_REPLY)?NULL:rpl, old_state, new_state);
+	if(dlg_run_event_route(dlg, (rpl==FAKED_REPLY)?NULL:rpl, old_state,
+			new_state)<0) {
+		/* dialog is gone */
+		return;
+	}
 
 	if (new_state==DLG_STATE_EARLY) {
 		run_dlg_callbacks(DLGCB_EARLY, dlg, req, rpl, DLG_DIR_UPSTREAM, 0);
@@ -1176,7 +1180,6 @@ dlg_cell_t *dlg_get_msg_dialog(sip_msg_t *msg)
 void dlg_onroute(struct sip_msg* req, str *route_params, void *param)
 {
 	dlg_cell_t *dlg = NULL;
-	dlg_cell_t *dlg0 = NULL;
 	dlg_iuid_t *iuid = NULL;
 	str val, callid, ftag, ttag;
 	int h_entry=0, h_id=0, new_state=0, old_state=0;
@@ -1315,15 +1318,9 @@ void dlg_onroute(struct sip_msg* req, str *route_params, void *param)
 	CURR_DLG_LIFETIME = (unsigned int)(time(0))-dlg->start_ts;
 	CURR_DLG_STATUS = new_state;
 
-	dlg_run_event_route(dlg, req, old_state, new_state);
-
-	dlg0 = dlg_lookup(h_entry, h_id);
-	if (dlg0==0) {
-		LM_ALERT("after event route - dialog not found [%u:%u] (%d/%d) (%p)\n",
-				h_entry, h_id, old_state, new_state, dlg);
+	if(dlg_run_event_route(dlg, req, old_state, new_state)<0) {
+		/* dialog is gone */
 		return;
-	} else {
-		dlg_release(dlg0);
 	}
 
 	/* delay deletion of dialog until transaction has died off in order
@@ -1519,7 +1516,10 @@ void dlg_ontimeout(struct dlg_tl *tl)
 		timeout_cb = (void *)CONFIRMED_DIALOG_STATE;
 	}
 
-	dlg_run_event_route(dlg, NULL, old_state, new_state);
+	if(dlg_run_event_route(dlg, NULL, old_state, new_state)<0) {
+		/* dialog is gone */
+		return;
+	}
 
 	if (new_state==DLG_STATE_DELETED && old_state!=DLG_STATE_DELETED) {
 		LM_WARN("timeout for dlg with CallID '%.*s' and tags '%.*s' '%.*s'\n",
@@ -1606,18 +1606,19 @@ int pv_get_dlg_status(struct sip_msg *msg, pv_param_t *param, pv_value_t *res)
 
 /*!
  * \brief Execute event routes based on new state
- *
+ * - returns: -1 if dialog doesn't exist after event route execution
+ *             0 if all ok
  */
-void dlg_run_event_route(dlg_cell_t *dlg, sip_msg_t *msg, int ostate, int nstate)
+int dlg_run_event_route(dlg_cell_t *dlg, sip_msg_t *msg, int ostate, int nstate)
 {
 	sip_msg_t *fmsg;
 	int rt;
 	int bkroute;
 
 	if(dlg==NULL)
-		return;
+		return -1;
 	if(ostate==nstate)
-		return;
+		return 0;
 
 	rt = -1;
 	if(nstate==DLG_STATE_CONFIRMED_NA) {
@@ -1630,7 +1631,7 @@ void dlg_run_event_route(dlg_cell_t *dlg, sip_msg_t *msg, int ostate, int nstate
 	}
 
 	if(rt==-1 || event_rt.rlist[rt]==NULL)
-		return;
+		return 0;
 
 	if(rt>=0 || dlg_event_callback.len>0) {
 		if(msg==NULL)
@@ -1639,7 +1640,7 @@ void dlg_run_event_route(dlg_cell_t *dlg, sip_msg_t *msg, int ostate, int nstate
 			fmsg = msg;
 
 		if (exec_pre_script_cb(fmsg, LOCAL_CB_TYPE)<=0)
-			return;
+			return 0;
 		dlg_ref(dlg, 1);
 		dlg_set_ctx_iuid(dlg);
 		LM_DBG("executing event_route %d on state %d\n", rt, nstate);
@@ -1651,6 +1652,7 @@ void dlg_run_event_route(dlg_cell_t *dlg, sip_msg_t *msg, int ostate, int nstate
 		dlg_unref(dlg, 1);
 		set_route_type(bkroute);
 	}
+	return 0;
 }
 
 int dlg_manage(sip_msg_t *msg)
diff --git a/modules/dialog/dlg_hash.h b/modules/dialog/dlg_hash.h
index f276f5c..19e774d 100644
--- a/modules/dialog/dlg_hash.h
+++ b/modules/dialog/dlg_hash.h
@@ -582,7 +582,7 @@ static inline int match_downstream_dialog(dlg_cell_t *dlg, str *callid, str *fta
 /*!
  *
  */
-void dlg_run_event_route(dlg_cell_t *dlg, sip_msg_t *msg, int ostate, int nstate);
+int dlg_run_event_route(dlg_cell_t *dlg, sip_msg_t *msg, int ostate, int nstate);
 
 
 int dlg_ka_add(dlg_cell_t *dlg);
diff --git a/modules/dialog/dlg_profile.c b/modules/dialog/dlg_profile.c
index 579d7d2..a692c35 100644
--- a/modules/dialog/dlg_profile.c
+++ b/modules/dialog/dlg_profile.c
@@ -437,6 +437,10 @@ int profile_cleanup( struct sip_msg *msg, unsigned int flags, void *param )
 {
 	dlg_cell_t *dlg;
 
+	if(get_route_type()==LOCAL_ROUTE) {
+		return 1;
+	}
+
 	current_dlg_msg_id = 0;
 	current_dlg_msg_pid = 0;
 	dlg = dlg_get_ctx_dialog();
diff --git a/modules/dialog/dlg_var.c b/modules/dialog/dlg_var.c
index 6884b9c..c251286 100644
--- a/modules/dialog/dlg_var.c
+++ b/modules/dialog/dlg_var.c
@@ -47,6 +47,9 @@ int msg_id;
 int dlg_cfg_cb(sip_msg_t *msg, unsigned int flags, void *cbp)
 {
 	dlg_cell_t *dlg;
+	if(get_route_type()==LOCAL_ROUTE) {
+		return 1;
+	}
 	if(flags&POST_SCRIPT_CB) {
 		dlg = dlg_get_ctx_dialog();
 		if(dlg!=NULL) {
@@ -76,6 +79,9 @@ int dlg_cfg_cb(sip_msg_t *msg, unsigned int flags, void *cbp)
 
 int cb_dlg_cfg_reset(sip_msg_t *msg, unsigned int flags, void *cbp)
 {
+	if(get_route_type()==LOCAL_ROUTE) {
+		return 1;
+	}
 	memset(&_dlg_ctx, 0, sizeof(dlg_ctx_t));
 
 	return 1;
@@ -83,6 +89,9 @@ int cb_dlg_cfg_reset(sip_msg_t *msg, unsigned int flags, void *cbp)
 
 int cb_dlg_locals_reset(sip_msg_t *msg, unsigned int flags, void *cbp)
 {
+	if(get_route_type()==LOCAL_ROUTE) {
+		return 1;
+	}
 	LM_DBG("resetting the local dialog shortcuts on script callback: %u\n", flags);
 	cb_dlg_cfg_reset(msg, flags, cbp);
 	cb_profile_reset(msg, flags, cbp);
@@ -918,6 +927,9 @@ dlg_ctx_t* dlg_get_dlg_ctx(void)
 
 int spiral_detect_reset(struct sip_msg *foo, unsigned int flags, void *bar)
 {
+	if(get_route_type()==LOCAL_ROUTE) {
+		return 1;
+	}
 	spiral_detected = -1;
 
 	return 0;
diff --git a/modules/ims_dialog/dlg_var.c b/modules/ims_dialog/dlg_var.c
index 323d7ef..0fac889 100644
--- a/modules/ims_dialog/dlg_var.c
+++ b/modules/ims_dialog/dlg_var.c
@@ -40,6 +40,9 @@ int msg_id;
 
 int dlg_cfg_cb(struct sip_msg *foo, unsigned int flags, void *bar)
 {
+	if(get_route_type()==LOCAL_ROUTE) {
+		return 1;
+	}
         memset(&_dlg_ctx, 0, sizeof(dlg_ctx_t));
 	return 1;
 }
